### PR TITLE
[Tripy] Disallow passing both Shapes and Tensors with the `convert_inputs_to_tensors decorator`

### PR DESCRIPTION
**Pull request by slyubomirsky**
Addresses issue https://github.com/NVIDIA/TensorRT-Incubator/issues/233, reopened from #265 and incorporating suggestions (thanks, @pranavm-nvidia). Per discussions, there are not many remaining uses of the `convert_inputs_to_tensors` decorator in the codebase and it also seems that in just about all of those cases, it would not make sense to accept both shapes and tensors.

### CODE REVIEW

File: tripy/tripy/frontend/utils.py [hunk 1]
```diff
@@ -245,26 +245,39 @@ def wrapper(*args, **kwargs):
...
-                (type(arg) for arg_name, arg in all_args if arg_name not in exclude),
-                key=lambda typ: {Tensor: 2, Shape: 1, ShapeScalar: 0}.get(typ, -1),
+            # Disallow mixing Tensor and Shape by default. If it makes sense in a given function
+            # to have both Tensor and Shape arguments, that might suggest that custom handling
+            # rather than relying on this decorator would make sense.
+            types = set(
+                {
+                    # There are other subclasses of Tensor, like Parameter and DefaultParameter.
+                    # Unless otherwise specified, we treat them as ordinary Tensors.
+                    Tensor if type(arg) != Shape and type(arg) != ShapeScalar else type(arg)
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/utils.py:255]:
```suggestion
                    Tensor if type(arg) not in {Shape, ShapeScalar} else type(arg)
```

File: tripy/tripy/frontend/utils.py [hunk 2]
```diff
@@ -245,26 +245,39 @@ def wrapper(*args, **kwargs):
...
-            # That is, if we have an operation with a mixture of types, all arguments are
-            # converted to the type with the highest precedence, e.g. Tensor + Shape -> Tensor.
-            TensorType = None
-            MaxType = max(
-                (type(arg) for arg_name, arg in all_args if arg_name not in exclude),
-                key=lambda typ: {Tensor: 2, Shape: 1, ShapeScalar: 0}.get(typ, -1),
+            # Disallow mixing Tensor and Shape by default. If it makes sense in a given function
+            # to have both Tensor and Shape arguments, that might suggest that custom handling
+            # rather than relying on this decorator would make sense.
+            types = set(
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/utils.py:251]: The `set()` here is not needed.

File: tripy/tripy/frontend/utils.py [hunk 3]
```diff
@@ -245,26 +245,39 @@ def wrapper(*args, **kwargs):
...
+                    for arg_name, arg in all_args
+                    if isinstance(arg, Tensor) and arg_name not in exclude
+                }
             )
-            if issubclass(MaxType, Tensor):
-                TensorType = MaxType
+            # We usually can treat ShapeScalars as either tensors or shapes due to broadcasting, so we can remove them from the below check.
+            shape_scalar_encountered = ShapeScalar in types
+            if ShapeScalar in types:
+                types.remove(ShapeScalar)
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/utils.py:263]: This is equivalent to:
```suggestion
            types -= {ShapeScalar}
```

Comment (slyubomirsky) [reply tripy/tripy/frontend/utils.py:263]: Good to know, I forgot to wrap the RHS in a set and was wondering why it didn't work.

File: tripy/tripy/frontend/utils.py (hunk 2 above)
Comment (slyubomirsky) [reply tripy/tripy/frontend/utils.py]: Removed, thanks.

**Review by pranavm-nvidia — APPROVED** [review]

**Pull request merged**